Modified iam_server_certificates_facts module to add path parameter

vijayanandsharma opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

Modified iam_server_certificates_facts module to include `Path` option. Improved module to use either name or path to retrieve the facts for IAM server certificates. 

Also refactored code to improve the module. If the specified certificates cannot be found, then the module return's an empty list.

<!--- If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
iam_server_certificates_facts 

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
2.7
```


##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful. -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

Commit:
Modified existing module to add path parameter and Refactored code to the new module standards

Commit:
Fixed pep8 Issues

ryansb commented on lib/ansible/modules/cloud/amazon/iam_server_certificate_facts.py:
```diff
@@ -1,10 +1,7 @@
-# Copyright: Ansible Project
-# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
-
-from __future__ import absolute_import, division, print_function
```
We can' t remove these `__future__` imports, they're important for python 2/3 compatibility.

ryansb commented on lib/ansible/modules/cloud/amazon/iam_server_certificate_facts.py:
```diff
@@ -1,10 +1,7 @@
-from __future__ import absolute_import, division, print_function
-__metaclass__ = type
 
+# Copyright: (c) 2018, Allen Sanabria (@linuxdynasty)
```
The copyright statement is set to `Ansible Project` for a reason, we maintain author information in the metadata instead.

ryansb commented on lib/ansible/modules/cloud/amazon/iam_server_certificate_facts.py:
```diff
@@ -81,89 +89,88 @@
-            }
+        if not name:
+            for server_cert in server_certs:
+                response['certificates'].append(connection.get_server_certificate(ServerCertificateName=server_cert['ServerCertificateName']))
```
This actually breaks the existing return format. The previous code creates a dictionary like:

```
{
    'MyCertName': { .... cert info object ...},
    'AnotherCertName': { .... cert info object ...},
}
```

The new code is instead creating a list, like:

```
[
     { .... cert info object ...},
     { .... cert info object ...},
]
```

We can't break backwards compatibility on the return of these facts.

wimnat replied:
@ryansb the module is marked as 'preview' so could we look at making this change?  It brings it in to line with other aws _facts modules by:

 - returning a list of dicts, with the cert name included in that dict
 - returning an empty list if not found rather than failing

Every other _facts module i can think of behaves this way.  I know it's a breaking change but i guess that's what the 'preview' type is for right?

vijayanandsharma replied:
@ryansb I agree with @wimnat. The existing module is not inlines with the other _facts module. 

Also please note that the existing module although contains the code to retrieve all server certificates, it is not used since `name` is marked as required. I have now added both `name` and `path` attribute to make this module more complete. 

With regards to the module output, I just tried to make it consistent with the other _facts modules.

ryansb replied:
I can see the value in it being a dictionary though, so you cloud do something like `cert_facts.certificates['www.foo.com']` instead of needing to search through the list. Is there a use case where the list is more useful? I think `cert_facts.certificates.values()` would be an ok way to get the list of certs

vijayanandsharma replied:
@ryansb, the use case here is to sort certs based on the upload date and use the most recent certificate when creating ELB's or ALB's. 

Assuming we need a certificate with a particular name, we can use the module with the `name` option which will return one list object that matches the `name`.

Another important reason for returning a list instead of the dict is to make this module consistent with the other _facts module which return a list of objects. For example: 

ec2_vpc_net_facts returns: 

`{ 
        "vpcs": [ ......
        ]
}`

ec2_vpc_subnet_facts returns: 

`{ 
        "subnets": [ ......
        ]
}`

wimnat replied:
It's true that it's useful to have a dict with the name as a key *if* that's what you're interested in.  But who's to say name should be the key?  Why not path, or certificate id?

With a list you can choose whatever you want using jinja selectattr.

E.g.

`certs.certificates | selectattr('server_certificate_metadata.server_certificate_name', 'equalto', 'www.foo.com') | list`
or if you wanted id
`certs.certificates | selectattr('server_certificate_metadata.server_certificate_id', 'equalto', 'xyz') | list`

This is how i select subnets in a particular AZ from the vpc subnet facts module.  I wouldn't expect a key of AZ because there's a possibility someone might want vpc_id or something else.

As for a use case for this module, I like this change because it would allow me to easily pick the cert that has the latest expiry date.  Something like:

```
"{{ certs.certificates | selectattr('server_certificate_metadata.path', 'equalto', '/') | sort(attribute='server_certificate_metadata.expiration') | last }}"
```

vijayanandsharma replied:
Hi @ryansb, 

Did you manage look at my comments? Appreciate your feedback on this.

Commit:
Fixed Review Comments and Removed Response Metadata from the module output

vijayanandsharma commented on lib/ansible/modules/cloud/amazon/iam_server_certificate_facts.py:
```diff
@@ -81,89 +89,88 @@
-            }
+        if not name:
+            for server_cert in server_certs:
+                response['certificates'].append(connection.get_server_certificate(ServerCertificateName=server_cert['ServerCertificateName']))
```
@ryansb I agree with @wimnat. The existing module is not inlines with the other _facts module. 

Also please note that the existing module although contains the code to retrieve all server certificates, it is not used since `name` is marked as required. I have now added both `name` and `path` attribute to make this module more complete. 

With regards to the module output, I just tried to make it consistent with the other _facts modules.

ryansb reviewed (changes requested):
Code looks good, but the return docs are now incorrect. Please write return docs reflecting the list format

Commit:
Updated Return Docs - Review Comments

Commit:
Fixed pep8 formatting issues

s-hertel commented on lib/ansible/modules/cloud/amazon/iam_server_certificate_facts.py:
```diff
@@ -35,134 +42,158 @@
+            response['certificates'].append(connection.get_server_certificate(ServerCertificateName=name))[
+                'ServerCertificate']
+        elif path:
+            server_certs = connection.list_server_certificates(PathPrefix=path)['ServerCertificateMetadataList']
```
It would be good if this used pagination, same below (https://boto3.readthedocs.io/en/latest/reference/services/iam.html#IAM.Paginator.ListServerCertificates).

s-hertel commented on lib/ansible/modules/cloud/amazon/iam_server_certificate_facts.py:
```diff
@@ -35,134 +42,158 @@
-        pass
+    except botocore.exceptions.ClientError as e:
+        if e.response['Error']['Code'] == 'NoSuchEntity':
+            return response['certificates']
```
Rather than having this and the call above in the same try/except, I think it would make sense to have a separate one for the `if not name:` block. That way, if name is not used and one of the server certificates that is initially listed is in the process of deleting this return may be premature (but it's right for the call on [line 138](https://github.com/ansible/ansible/pull/42609/files#diff-c7c2a98372cabb503e540623363950b8R138)).

Maybe something along the lines of

```
if not name:
    for server_cert in server_certs:
        try:
            # get_server_certificate and append
        except ClientError as e:
            if e.response['Error']['Code'] == 'NoSuchEntity':
                continue  # following server_certs may still exist
```

s-hertel reviewed (commented):
A couple minor things, but looks good to me.

Commit:
Implemented Paginator to List Server Certificates

vijayanandsharma commented:
Hi @wimnat @ryansb I'm happy to be the maintainer for the module. Please let me know if this can be merged to next version.

vijayanandsharma committed:
Rebased branch and updated the module name

vijayanandsharma committed:
Delete iam_server_certificate_facts.py as symlink to info module already exists

vijayanandsharma committed:
Fixing sanity test failures

vijayanandsharma committed:
Removed iam_server_certificate_info from the ignore file

vijayanandsharma committed:
Updated versions added

tremble commented on lib/ansible/modules/cloud/amazon/iam_server_certificate_info.py:
```diff
@@ -10,22 +10,28 @@
 short_description: Retrieve the information of a server certificate
 description:
-  - Retrieve the attributes of a server certificate.
+  - Retrieve the attributes of a server certificate
```
```suggestion
  - Retrieve the attributes of a server certificate.
```

Description fields should be complete sentences.

tremble replied:
See also https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#documentation-block

tremble commented on lib/ansible/modules/cloud/amazon/iam_server_certificate_info.py:
```diff
@@ -10,22 +10,28 @@
   - This module was called C(iam_server_certificate_facts) before Ansible 2.9. The usage did not change.
 version_added: "2.2"
-author: "Allen Sanabria (@linuxdynasty)"
+author: "Vijayanand Sharma (@vijayanandsharma)"
```
```suggestion
author:
- "Allen Sanabria (@linuxdynasty)"
- "Vijayanand Sharma (@vijayanandsharma)"
```
While you've rewritten much of the code it's obviously not a clean-room implementation.  Removing all of @linuxdynasty's credit here is somewhat dubious.

jillr replied:
Module authorship also defines maintainers for the module, so this removal would have a functional effect.  Unless they themselves elect to step down we should definitely keep linuxdynasty here.

tremble commented on lib/ansible/modules/cloud/amazon/iam_server_certificate_info.py:
```diff
@@ -10,22 +10,28 @@
   name:
     description:
-      - The name of the server certificate you are retrieving attributes for.
+      - The name of the server certificate you are retrieving attributes for. This argument is mutually exclusive with I(path)
```
```suggestion
      - The name of the server certificate you are retrieving attributes for.
      - This argument is mutually exclusive with I(path).
```

1. Rather than creating a long line, this can be split into two lines.
2. Please use full sentences (including a full stop) for descriptions.

tremble commented on lib/ansible/modules/cloud/amazon/iam_server_certificate_info.py:
```diff
@@ -10,22 +10,28 @@
+    required: false
+  path:
+    description:
+      - The path to the server certificate you are retrieving attributes for. This argument is mutually exclusive with I(name)
```
```suggestion
      - The path to the server certificate you are retrieving attributes for.
      - This argument is mutually exclusive with I(name).
```

tremble commented on lib/ansible/modules/cloud/amazon/iam_server_certificate_info.py:
```diff
@@ -36,136 +42,166 @@
+            type: complex
+            contains:
+                server_certificate_id:
+                    description: The 21 character certificate id
```
```suggestion
                    description: The 21 character certificate id.
```

tremble commented on lib/ansible/modules/cloud/amazon/iam_server_certificate_info.py:
```diff
@@ -36,136 +42,166 @@
+                    type: str
+                    sample: "ADWAJXWTZAXIPIMQHMJPO"
+                server_certificate_name:
+                    description: The name of the server certificate
```
```suggestion
                    description: The name of the server certificate.
```

tremble commented on lib/ansible/modules/cloud/amazon/iam_server_certificate_info.py:
```diff
@@ -36,136 +42,166 @@
+                    type: str
+                    sample: "server-cert-name"
+                arn:
+                    description: The Amazon resource name of the server certificate
```
```suggestion
                    description: The Amazon resource name of the server certificate.
```

tremble commented on lib/ansible/modules/cloud/amazon/iam_server_certificate_info.py:
```diff
@@ -36,136 +42,166 @@
+                    type: str
+                    sample: "arn:aws:iam::911277865346:server-certificate/server-cert-name"
+                path:
+                    description: The path of the server certificate
```
```suggestion
                    description: The path of the server certificate.
```

tremble commented on lib/ansible/modules/cloud/amazon/iam_server_certificate_info.py:
```diff
@@ -36,136 +42,166 @@
+
+    if name:
+        try:
+            response['certificates'].append(connection.get_server_certificate(ServerCertificateName=name)['ServerCertificate'])
```
Since you're already using the AWSRetry decorator, it's worth wrapping this call too.  (The big one iIt would catch is API Rate Limit errors.)
If you include the the NoSuchEntity handling, make sure you rethrow any other ClientError exceptions.  Avoid handling any other exceptions inside your wrapper as that would prevent the decorator from doing its job.

tremble commented on lib/ansible/modules/cloud/amazon/iam_server_certificate_info.py:
```diff
@@ -36,136 +42,166 @@
-            }
+            try:
+                response['certificates'].append(
+                    connection.get_server_certificate(ServerCertificateName=server_cert['ServerCertificateName'])['ServerCertificate'])
```
Again, here I'd recommend wrapping this with the AWSRetry decorator.

tremble commented on lib/ansible/modules/cloud/amazon/iam_server_certificate_info.py:
```diff
@@ -36,136 +42,166 @@
+            try:
+                response['certificates'].append(
+                    connection.get_server_certificate(ServerCertificateName=server_cert['ServerCertificateName'])['ServerCertificate'])
+            except botocore.exceptions.ClientError as e:
```
Did you mean to simply drop any other ClientErrors on the floor here?  (You're also not handling BotoCoreError this time around.)

vijayanandsharma replied:
I have handled this now. Thank you for pointing this out

tremble commented on lib/ansible/modules/cloud/amazon/iam_server_certificate_info.py:
```diff
@@ -36,136 +42,166 @@
-    cert_name = module.params.get('name')
-    results = get_server_certs(iam, cert_name)
-    module.exit_json(results=results)
+    argument_spec = (
```
There's an extra set of parentheses here.  This only needs 
```
argument_spec = dict(...)
```

tremble commented on lib/ansible/modules/cloud/amazon/iam_server_certificate_info.py:
```diff
@@ -36,136 +42,166 @@
+    paginator = connection.get_paginator('list_server_certificates')
+    if path:
+        return paginator.paginate(PathPrefix=path).build_full_result()
+    else:
```
Since you always return in your if statement, this else isn't technically required.

```
if path:
  return ....

return ....
```

tremble reviewed (changes requested):
Mostly little niggles, but I really don't like the complete removal (by anyone other @linuxdynasty) of @linuxdynasty from the Author list.

jillr reviewed (commented):
This PR changes the structure and keys of the data being returned, which could break things for users.  We should be able to use a deprecation period to allow people time to transition to the new keys by returning the old dictionary under `ansible_module_results` as well as the new results.
For the new results though, is it necessary to nest values under `certificates`?  This module should only ever return results for a single certificate, is that correct?

vijayanandsharma committed:
Fixed Review Comments

vijayanandsharma commented:
@tremble @jillr, 

Thanks you for your feedback, really appreciate it. I have addressed all of your comments now, could you please review the PR again?

@jillr with regards your comments, please find my reply below: 

> This PR changes the structure and keys of the data being returned, which could break things for users. We should be able to use a deprecation period to allow people time to transition to the new keys by returning the old dictionary under `ansible_module_results` as well as the new results.
> For the new results though, is it necessary to nest values under `certificates`? This module should only ever return results for a single certificate, is that correct?

Well the facts module may return one or more certificates based on the arguments passed to the module i.e it will return a list of 1 object if you specify the name, but it would return multiple certificates if you specify path. And hence this is nested under certificates asn

vijayanandsharma committed:
Fixed PEP8 Issue

vijayanandsharma committed:
File deleted in devel but got added because of the rebase - Deleted files
